functorch/_src/aot_autograd.py

```diff
@@ -441,7 +472,45 @@ def backward(ctx, *flat_args):
     def compiled_function(*args):
         return CompiledFunction.apply(*remove_dupe_args(args))
 
-    return compiled_function
+    # Note: [Calling Convention for AOTAUtograd forward/backward]
+    # The function returned by AOTAutograd has a few moving parts
+    # that require an internal calling convention.
+    # In order to handle input mutations, the function that AOTAutograd returns
+    # consists of:
+    # - A purely functional Autograd.Function
+    # - An opaque epilogue that runs after the forward, performing any input mutations.
+    #
+    # Let's say we have a function that takes in N inputs, returns M outputs,
+    # and also performs input mutations on K of its N inputs.
+    # The calling convention is:
+    #
+    # CompiledFunction.forward(inputs) -> mutated_inputs, outputs
+    # CompiledFunction.backward(grad_outputs) -> grad_inputs
+    # Epilogue(inp1, mutated_inp1, inp2, mutated_inp2, ...) -> ()
+    #
+    # TODO: Doesn't aot_dispatch_base need the epilogue too?
+    # ... it probably needs to run functionalization as well?
+    @disable_torchdynamo
+    def compiled_function_with_epilogue(*args):
+        # First, call the custom autograd fn
+        fw_outs = compiled_function(*args)
+
+        # Then, perform the opaque epilogue to mutate any inputs.
+        # This needs to happen *outside** of the autograd.function,
+        # because torch.autograd.function can't handle input mutations in some cases.
+        if config.use_functionalize:
+            # Call the input mutation submodule with the mutated inputs
+            mutated_inputs = fw_outs[:_num_mutated_inputs]
+            fw_outs = fw_outs[_num_mutated_inputs:]
+            # The input mutation epilogue expects inputs in the same order as the original graph,
+            # but in the order (inpt1_old, inpt1_new, inpt2_old, inpt2_new, ...)
+            mutated_input_args = [x for pair in zip(original_inputs_needing_mutation, mutated_inputs) for x in pair]
+            # TODO: this epilogue should also be responsible for generating outputs
+            # that are aliases of inputs.
+            input_mutation_epilogue(*mutated_input_args)
```

Comment:
So after the pass, all mutation would be in opaque mutation epilogue and backends lose visibility there.
We are missing out on fusion opportunities here.

Would we be able to opt-in to inline the mutation epilogue back into the main graph?

Reply:
> So after the pass, all mutation would be in opaque mutation epilogue

TBC, this is only for captured graphs with input mutations. Intermediate mutations in a graph would have already been removed by functionalization.

The idea in this PR as its stands is that instead of the backend seeing an operator like:
```
native_batch_norm(Tensor input, Tensor? weight, Tensor? bias, Tensor(a!) running_mean, Tensor(a!) running_var, bool training, float momentum, float eps) -> (Tensor, Tensor, Tensor)
```

It'll see a purely functional version, that returns the updated inputs instead of mutating them directly:
```
native_batch_norm(Tensor input, Tensor? weight, Tensor? bias, Tensor running_mean, Tensor running_var, bool training, float momentum, float eps) -> (Tensor, Tensor, Tensor, Tensor, Tensor)
```

And if mutated inputs to that operator happen to correspond to graph inputs (which is true for running_mean/running_var variables for batch norm), then at the end of the graph we'll have an epilogue that copies the "updated inputs" back to the original inputs:
```
def compiled_fn(inpt1, inpt2):
     # ... first run the entirely functional compiled function
     outs, mutated_inpt1, mutated_inpt2 = real_compiled_fn(inpt1, inpt2)
     inpt1.copy_(mutated_inpt1)
     inpt2.copy_(mutated_inpt2)
     return outs
```

> Would we be able to opt-in to inline the mutation epilogue back into the main graph?

I remember @Chillee bringing this up before - we probably can? Although for now, this PR doesn't do that and just ensures that we do the "correct" thing first.

It's also not clear to me - why would this prevent fusions? The main disadvantage as I see it is that you can end up using more memory - you have to keep the buffer for both the original input and the updated input around while the compiled function is running.

Reply:
Thanks for the explanation, I think we are mostly on the same page here. I'm mostly asking for mutation on inputs, since that's what normalization layers uses for running stats update.

> It's also not clear to me - why would this prevent fusions?

It's the epilogue in-place copies that we are missing out. Since those are cheap and easy to handle in normalization kernels. i.e. If we keep those in the epilogue, they won't be visible to fuser backend.